File: scripts/find_splits.py
"""
Find all PositionSplit transactions on Polymarket's Conditional Tokens contract
in the last 24 hours. Count per account and sum USDC amounts.

Handles Etherscan's 10K result limit by chunking the block range.
Separates CTF Exchange (orderbook fills) from direct user splits.
"""
import json
import time
import urllib.request
from collections import defaultdict

API_KEY = "C5MMHWYARZTRK8YVVCM9DAUU5ZIASC2CKN"
CT_CONTRACT = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045"
TOPIC0 = "0x2e6bb91f8cbcda0c93623c54d0403a43514fabc40084ec96b6d5379a74786298"
USDC_E = "0x2791bca1f2de4661ed88a30c99a7a9449aa84174"

# Known Polymarket contracts (not real user wallets)
POLYMARKET_CONTRACTS = {
    "0x4bfb41d5b3570defd03c39a9a4d8de6bd8b8982e",  # CTF Exchange
    "0xc5d563a36ae78145c45a50134d48a1215220f80a",  # NegRisk CTF Exchange
    "0x78769d50be1763ed1ca0d5e878d93f05aabff29e",  # NegRisk Fee Module
    "0x6a9d222616c90fca5754cd1333cfd9b7fb6a4f74",  # UmaCtf Adapter
}

BASE_URL = "https://api.etherscan.io/v2/api"


def api_call(url, retries=3):
    for attempt in range(retries):
        try:
            resp = json.loads(urllib.request.urlopen(url, timeout=30).read())
            return resp
        except Exception as e:
            if attempt < retries - 1:
                time.sleep(1)
            else:
                raise


def get_current_block():
    url = f"{BASE_URL}?chainid=137&module=proxy&action=eth_blockNumber&apikey={API_KEY}"
    resp = api_call(url)
    return int(resp["result"], 16)


def get_logs(from_block, to_block, page=1, offset=1000):
    url = (
        f"{BASE_URL}?chainid=137&module=logs&action=getLogs"
        f"&address={CT_CONTRACT}&topic0={TOPIC0}"
        f"&fromBlock={from_block}&toBlock={to_block}"
        f"&page={page}&offset={offset}"
        f"&apikey={API_KEY}"
    )
    resp = api_call(url)
    if resp["status"] == "1":
        return resp["result"]
    return []


def get_tx_sender(tx_hash):
    """Get the actual sender (from) of a transaction."""
    url = (
        f"{BASE_URL}?chainid=137&module=proxy&action=eth_getTransactionByHash"
        f"&txhash={tx_hash}&apikey={API_KEY}"
    )
    resp = api_call(url)
    result = resp.get("result")
    if result:
        return result.get("from", "").lower()
    return None


def get_all_logs(from_block, to_block):
    """Fetch all logs, chunking the block range to avoid 10K limit."""
    all_logs = []
    chunk_size = (to_block - from_block) // 4  # Start with 4 chunks

    ranges = []
    start = from_block
    while start < to_block:
        end = min(start + chunk_size, to_block)
        ranges.append((start, end))
        start = end + 1

    for i, (r_from, r_to) in enumerate(ranges):
        page = 1
        chunk_total = 0
        while True:
            logs = get_logs(r_from, r_to, page=page, offset=1000)
            if not logs:
                break
            all_logs.extend(logs)
            chunk_total += len(logs)
            if len(logs) < 1000:
                break
            page += 1
            time.sleep(0.22)

        print(f"  Chunk {i+1}/{len(ranges)} (blocks {r_from}-{r_to}): {chunk_total} events")
        time.sleep(0.22)

    return all_logs


def parse_log(log):
    stakeholder = "0x" + log["topics"][1][-40:]
    data = log["data"]
    d = data[2:]

    collateral_token = "0x" + d[24:64]

    amount_hex = d[128:192]
    amount = int(amount_hex, 16)

    is_usdc = collateral_token.lower() == USDC_E.lower()
    usdc_amount = amount / 1e6 if is_usdc else 0

    tx_hash = log["transactionHash"]
    timestamp = int(log["timeStamp"], 16)
    block = int(log["blockNumber"], 16)

    return {
        "stakeholder": stakeholder.lower(),
        "collateral_token": collateral_token.lower(),
        "amount_raw": amount,
        "usdc_amount": usdc_amount,
        "is_usdc": is_usdc,
        "tx_hash": tx_hash,
        "timestamp": timestamp,
        "block": block,
    }


def main():
    current_block = get_current_block()
    blocks_24h = int(86400 / 2.1)
    from_block = current_block - blocks_24h

    print(f"Current block: {current_block}")
    print(f"24h ago block: {from_block}")
    print(f"Scanning PositionSplit events (chunked)...\n")

    all_logs = get_all_logs(from_block, current_block)
    print(f"\nTotal PositionSplit events: {len(all_logs)}")

    # Parse all events
    parsed = []
    non_usdc = 0
    for log in all_logs:
        try:
            p = parse_log(log)
            if p["is_usdc"]:
                parsed.append(p)
            else:
                non_usdc += 1
        except Exception as e:
            print(f"  Parse error: {e}")

    if non_usdc > 0:
        print(f"  ({non_usdc} non-USDC splits skipped)")
    print(f"USDC split events: {len(parsed)}")

    # Separate exchange splits from direct user splits
    exchange_splits = [p for p in parsed if p["stakeholder"] in POLYMARKET_CONTRACTS]
    direct_splits = [p for p in parsed if p["stakeholder"] not in POLYMARKET_CONTRACTS]

    # --- DIRECT USER SPLITS ---
    direct_accounts = defaultdict(lambda: {"count": 0, "total_usdc": 0.0, "tx_hashes": set()})
    for p in direct_splits:
        acc = direct_accounts[p["stakeholder"]]
        acc["count"] += 1
        acc["total_usdc"] += p["usdc_amount"]
        acc["tx_hashes"].add(p["tx_hash"])

    # --- EXCHANGE SPLITS: look up tx senders for unique tx hashes ---
    exchange_tx_hashes = set(p["tx_hash"] for p in exchange_splits)
    print(f"\nExchange-mediated splits: {len(exchange_splits)} events in {len(exchange_tx_hashes)} txns")
    print(f"Direct user splits: {len(direct_splits)} events")

    # For exchange splits, look up actual tx senders (sample up to 200 txns to save API calls)
    exchange_accounts = defaultdict(lambda: {"count": 0, "total_usdc": 0.0, "tx_hashes": set()})
    tx_to_sender = {}

    sorted_exchange_txs = list(exchange_tx_hashes)
    # Group exchange splits by tx_hash for amount lookup
    tx_to_amount = defaultdict(float)
    tx_to_count = defaultdict(int)
    for p in exchange_splits:
        tx_to_amount[p["tx_hash"]] += p["usdc_amount"]
        tx_to_count[p["tx_hash"]] += 1

    print(f"\nLooking up senders for {min(len(sorted_exchange_txs), 500)} exchange txns...")
    looked_up = 0
    for tx_hash in sorted_exchange_txs[:500]:
        sender = get_tx_sender(tx_hash)
        if sender:
            tx_to_sender[tx_hash] = sender
            acc = exchange_accounts[sender]
            acc["count"] += tx_to_count[tx_hash]
            acc["total_usdc"] += tx_to_amount[tx_hash]
            acc["tx_hashes"].add(tx_hash)
        looked_up += 1
        if looked_up % 50 == 0:
            print(f"  ...looked up {looked_up}/{min(len(sorted_exchange_txs), 500)} txns")
        time.sleep(0.22)

    # Remaining txns we couldn't look up
    remaining_txs = sorted_exchange_txs[500:]
    remaining_usdc = sum(tx_to_amount[tx] for tx in remaining_txs)
    remaining_count = sum(tx_to_count[tx] for tx in remaining_txs)

    # --- PRINT RESULTS ---
    print(f"\n{'='*90}")
    print(f"POLYMARKET SPLIT SUMMARY (last 24h)")
    print(f"{'='*90}")

    total_usdc_all = sum(p["usdc_amount"] for p in parsed)
    total_txs_all = len(set(p["tx_hash"] for p in parsed))
    print(f"Total split events:     {len(parsed)}")
    print(f"Total unique txns:      {total_txs_all}")
    print(f"Total USDC split:       ${total_usdc_all:,.2f}")
    print(f"  via Exchange:         ${sum(p['usdc_amount'] for p in exchange_splits):,.2f} ({len(exchange_splits)} events)")
    print(f"  direct:               ${sum(p['usdc_amount'] for p in direct_splits):,.2f} ({len(direct_splits)} events)")

    # Print exchange users (actual tx senders)
    sorted_ex = sorted(exchange_accounts.items(), key=lambda x: x[1]["total_usdc"], reverse=True)
    print(f"\n--- EXCHANGE USERS (tx senders, top 50) ---")
    print(f"{'Address':<44} {'Splits':>7} {'Txns':>6} {'USDC Amount':>15}")
    print("-" * 80)
    for addr, data in sorted_ex[:50]:
        label = ""
        if addr in POLYMARKET_CONTRACTS:
            label = " [CONTRACT]"
        print(f"{addr}{label}  {data['count']:>7}  {len(data['tx_hashes']):>5}  ${data['total_usdc']:>14,.2f}")
    if len(sorted_ex) > 50:
        print(f"  ... and {len(sorted_ex) - 50} more accounts")
    if remaining_txs:
        print(f"  (+ {len(remaining_txs)} txns not looked up: ~${remaining_usdc:,.2f})")

    # Print direct splitters
    sorted_direct = sorted(direct_accounts.items(), key=lambda x: x[1]["total_usdc"], reverse=True)
    if sorted_direct:
        print(f"\n--- DIRECT SPLITTERS (called splitPosition directly) ---")
        print(f"{'Address':<44} {'Splits':>7} {'Txns':>6} {'USDC Amount':>15}")
        print("-" * 80)
        for addr, data in sorted_direct[:30]:
            print(f"{addr}  {data['count']:>7}  {len(data['tx_hashes']):>5}  ${data['total_usdc']:>14,.2f}")

    # Combined unique user count
    all_users = set(exchange_accounts.keys()) | set(direct_accounts.keys())
    all_users -= POLYMARKET_CONTRACTS
    print(f"\n{'='*90}")
    print(f"UNIQUE USER ACCOUNTS: {len(all_users)}")
    print(f"{'='*90}")


if __name__ == "__main__":
    main()

File: scripts/find_splits.mjs
/**
 * Find all Polymarket accounts that performed "split" transactions
 * (splitPosition on CTF contract) in the last 24 hours.
 *
 * Usage: node scripts/find_splits.mjs
 *
 * Zero dependencies — Node.js 18+ built-ins only.
 * Uses Polygon RPC (eth_getLogs) with incremental aggregation.
 */

const CTF_ADDRESS = '0x4d97dcd97ec945f40cf65f87097ace5ea0476045';
const POSITION_SPLIT_TOPIC = '0x2e6bb91f8cbcda0c93623c54d0403a43514fabc40084ec96b6d5379a74786298';
const ETHERSCAN_API_KEY = 'C5MMHWYARZTRK8YVVCM9DAUU5ZIASC2CKN';
const ETHERSCAN_BASE = 'https://api.etherscan.io/v2/api';
const RPC_URL = 'https://polygon.drpc.org';
const CHUNK_BLOCKS = 400; // ~31K events per chunk, safely under drpc timeout
const RPC_DELAY_MS = 2000; // generous delay for free RPC
const USDC_DECIMALS = 6;

const sleep = (ms) => new Promise(r => setTimeout(r, ms));

// ── Etherscan (block timestamp only) ─────────────────────────────────

async function getBlockByTimestamp(timestamp) {
  for (let attempt = 0; attempt < 5; attempt++) {
    await sleep(220);
    const qs = new URLSearchParams({
      chainid: '137',
      module: 'block',
      action: 'getblocknobytime',
      timestamp: String(timestamp),
      closest: 'before',
      apikey: ETHERSCAN_API_KEY,
    });
    try {
      const res = await fetch(`${ETHERSCAN_BASE}?${qs}`, { signal: AbortSignal.timeout(15000) });
      const json = await res.json();
      if (json.status === '1') return parseInt(json.result, 10);
    } catch {}
    await sleep(2000);
  }
  throw new Error('Could not resolve block by timestamp');
}

// ── Polygon RPC ──────────────────────────────────────────────────────

let rpcId = 0;

async function rpcGetLogs(fromBlock, toBlock) {
  const body = JSON.stringify({
    jsonrpc: '2.0',
    id: ++rpcId,
    method: 'eth_getLogs',
    params: [{
      address: CTF_ADDRESS,
      topics: [POSITION_SPLIT_TOPIC],
      fromBlock: '0x' + fromBlock.toString(16),
      toBlock: '0x' + toBlock.toString(16),
    }],
  });

  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const res = await fetch(RPC_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Connection: 'close' },
        body,
        signal: AbortSignal.timeout(120000),
      });
      const json = await res.json();

      if (json.error) {
        const msg = json.error.message || JSON.stringify(json.error);
        if (msg.includes('too many') || msg.includes('more than') || msg.includes('exceed') || msg.includes('block range')) {
          return { tooMany: true };
        }
        throw new Error(msg);
      }
      return { logs: json.result || [] };
    } catch (err) {
      if (err.message?.includes('too many')) return { tooMany: true };
      if (attempt < 4) {
        await sleep((attempt + 1) * 3000);
        continue;
      }
      throw err;
    }
  }
}

// ── Incremental processing ───────────────────────────────────────────

const accounts = new Map();
let totalEvents = 0;

function processLogs(logs) {
  for (const log of logs) {
    const stakeholder = '0x' + log.topics[1].slice(26).toLowerCase();
    const dataHex = log.data.slice(2);
    const amountHex = dataHex.slice(128, 192);
    const amount = BigInt('0x' + amountHex);

    const existing = accounts.get(stakeholder);
    if (existing) {
      existing.count++;
      existing.totalAmount += amount;
    } else {
      accounts.set(stakeholder, { count: 1, totalAmount: amount });
    }
    totalEvents++;
  }
}

// ── Main ─────────────────────────────────────────────────────────────

function formatUSDC(amountBigInt) {
  const str = amountBigInt.toString().padStart(USDC_DECIMALS + 1, '0');
  const whole = str.slice(0, -USDC_DECIMALS) || '0';
  const frac = str.slice(-USDC_DECIMALS);
  return whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',') + '.' + frac.slice(0, 2);
}

async function main() {
  console.log('Fetching PositionSplit events from last 24 hours...\n');

  const now = Math.floor(Date.now() / 1000);
  const yesterday = now - 86400;
  console.log(`  From: ${new Date(yesterday * 1000).toISOString()}`);
  console.log(`  To:   ${new Date(now * 1000).toISOString()}`);

  const fromBlock = await getBlockByTimestamp(yesterday);
  const toBlock = await getBlockByTimestamp(now);
  const totalBlocks = toBlock - fromBlock;
  console.log(`  Blocks: ${fromBlock} -> ${toBlock} (~${totalBlocks} blocks)\n`);

  const numChunks = Math.ceil(totalBlocks / CHUNK_BLOCKS);
  let apiCalls = 0;
  const startTime = Date.now();

  for (let i = 0; i < numChunks; i++) {
    const chunkFrom = fromBlock + i * CHUNK_BLOCKS;
    const chunkTo = Math.min(chunkFrom + CHUNK_BLOCKS - 1, toBlock);

    await sleep(RPC_DELAY_MS);
    const result = await rpcGetLogs(chunkFrom, chunkTo);
    apiCalls++;

    if (result.tooMany) {
      // Split this chunk in half
      const mid = Math.floor((chunkFrom + chunkTo) / 2);
      await sleep(RPC_DELAY_MS);
      const left = await rpcGetLogs(chunkFrom, mid);
      apiCalls++;
      if (left.logs) processLogs(left.logs);

      await sleep(RPC_DELAY_MS);
      const right = await rpcGetLogs(mid + 1, chunkTo);
      apiCalls++;
      if (right.logs) processLogs(right.logs);
    } else {
      processLogs(result.logs);
    }

    const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
    const rate = (i + 1) > 0 ? (Date.now() - startTime) / (i + 1) : 1000;
    const eta = ((numChunks - i - 1) * rate / 1000).toFixed(0);
    process.stdout.write(
      `\r  ${i + 1}/${numChunks} chunks | ${totalEvents.toLocaleString()} events | ${accounts.size} accounts | ${elapsed}s / ~${eta}s left`
    );
  }

  console.log('\n');

  if (totalEvents === 0) {
    console.log('No split events found.');
    return;
  }

  // Sort and print
  const sorted = [...accounts.entries()]
    .map(([address, { count, totalAmount }]) => ({ address, count, totalAmount }))
    .sort((a, b) => (b.totalAmount > a.totalAmount ? 1 : b.totalAmount < a.totalAmount ? -1 : 0));

  console.log('='.repeat(78));
  console.log('  Polymarket Split Accounts (Last 24 Hours)');
  console.log('='.repeat(78));
  console.log('');
  console.log(
    '  #'.padEnd(7) +
    'Address'.padEnd(44) +
    'Splits'.padStart(8) +
    'Total USDC'.padStart(18)
  );
  console.log('-'.repeat(78));

  for (let i = 0; i < sorted.length; i++) {
    const { address, count, totalAmount } = sorted[i];
    console.log(
      `  ${i + 1}.`.padEnd(7) +
      address.padEnd(44) +
      String(count).padStart(8) +
      ('$' + formatUSDC(totalAmount)).padStart(18)
    );
  }

  console.log('');
  console.log('='.repeat(78));
  const totalSplits = sorted.reduce((s, r) => s + r.count, 0);
  const totalUSDC = sorted.reduce((s, r) => s + r.totalAmount, 0n);
  console.log(`  Unique accounts: ${sorted.length}`);
  console.log(`  Total splits:    ${totalSplits.toLocaleString()}`);
  console.log(`  Total USDC:      $${formatUSDC(totalUSDC)}`);
  console.log(`  API calls:       ${apiCalls}`);
  console.log(`  Time:            ${((Date.now() - startTime) / 1000).toFixed(1)}s`);
  console.log('='.repeat(78));
}

main().catch(err => {
  console.error('\nFatal:', err.message || err);
  process.exit(1);
});
